### DISCUSSION TITLE
Alternative Directory Names

### DISCUSSION THREAD

**Discussion by pi0** [pi0/config-dir#3] [General]
> [!NOTE]
> Here are some alternative ideas for the config directory name collected from several places.
>
> See [previous efforts](https://github.com/pi0/config-dir/discussions/5) and [current tools](https://github.com/pi0/config-dir/discussions/6) for other relevant dicussions.

### `.config/` (current proposal)

Config directory indicates it holds the configuration files by their name and is a refactored part of the common `.config` suffix in current config file names.

It is prefixed with a dot `.` so that it will be sorted as the first directory and more clearly separated from the rest of the project, reducing more noise. Also is consistent with [XDG](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html)'s `~/.config` directory. 

### `.conf/`

Before writing this proposal I had been thinking of `.conf/` as a shorter alternative. As a drawback, this name is less readable.

### `.meta/`

[@barelyreaper](https://twitter.com/barelyreaper) suggested in a [tweet](https://twitter.com/barelyreaper/status/1757385448266355025) to use `.meta/` since most of the (current) config files have `.config` in their name and with a more generic place like `.meta/` we can put other kinds of dotfiles such as `prettierignore` in it. 

### `config/`

[@nainemom](https://github.com/nainemom) @pawamoy suggested using the `config` directory instead of `.config` so that it won't be hidden in Unix tools. ([#1](https://github.com/pi0/config-dir/issues/1), [#13](https://github.com/pi0/config-dir/issues/13))

### `package.json`

[@jullerino](https://twitter.com/jullerino) suggested in a [tweet](https://twitter.com/remcohaszing/status/1757488879362310406) to prefer `package.json` that is also useful for monorepos. [@remcohaszing](https://twitter.com/remcohaszing/) has some [interesting points](https://twitter.com/remcohaszing/status/1757488651812897131) why this might not always work.

**Comment by pawamoy** [comment]
In one of my projects, years ago, I added support for a `.configconfig` file which contained a single line with a directory name, in which we would find other config files. I named it `.configconfig` in reference to `.editorconfig`... :joy:

**Comment by adamdehaven** [comment]
Have you considered `/.configs/` rather than `/.config/`? (essentially, make it plural)

- The `/.config/` directory is _already_ fairly standardized for system-level configurations as well as app-specific write locations on Unix/Linux systems (e.g. `~/.config/` on your Mac). 
  - Making it "different enough" can prevent downstream issues, e.g. if some systems restrict access to directories based on the name (looking at you Netlify)
- It holds multiple "configurations," not a single "config" :trollface:

**Reply by joshtriplett** [reply]
See https://github.com/pi0/config-dir/discussions/3#discussioncomment-15419197 - it holds "configuration", not "configurations".

**Comment by phanect** [comment]
I'm delayed, but I suggest `_configs/` (or `_config/` if singular is preferable)

Reasons:

# 1. Files with a dot prefix in the name: They are treated as hidden files on Linux (and maybe Mac)

`.config/` and other directory names starting with a dot are treated as hidden directories on Linux and other Unix(-like) systems (maybe including Mac?), as mentioned in the suggestion for `config`.

This is one of the reasons why the Vite team didn't accept this .config/ proposal.

https://github.com/vitejs/vite/issues/15911#issuecomment-2076259187

`.config/`, `.conf/`, and `.meta/` have this problem.
However, `_conf/`, and `_meta/` work as the alternatives.

# 2. `config/` conflicts with some frameworks (especially PHP's)

`config/` does not work in some cases because some frameworks use `config/` as its configuration directory and the directory name is overlapped.
I have never seen such frameworks using the `config/` directory in JS ecosystems, but some PHP frameworks use it. The following PHP frameworks use `config/` as their configuration directory.

- [Laravel](https://laravel.com/docs/11.x/structure#the-config-directory)
- [Symfony](https://symfony.com/doc/current/configuration.html)
- [CakePHP](https://book.cakephp.org/5/en/development/configuration.html#configuring-your-application)

If the scope of this proposal is not only JavaScript and TypeScript projects, I think this should be considered.
Even if this proposal only targets JS/TS, it might cause trouble if the developer mixes their PHP and frontend code in the same repository.

# 3. package.json: causes more conflicts

In addition to remcohaszing's points, I also heard an opinion to avoid writing configs in package.json.

If you write all the configs in package.json in the team projects, it causes a lot of Git conflicts because many people frequently update package.json.

**Reply by epage** [reply]
imo some project configuration is too important to hide for users by default.  For this reason, my tool `typos` offers the user the choice `typos.toml`, `.typos.toml`, or `_typos.toml`.

If this is a general effort, I feel like standardizing on `_config/` would be better than `.config/`

**Reply by phanect** [reply]
As I replied in https://github.com/pi0/config-dir/discussions/3#discussioncomment-12790245 and mentioned in its parent thread, .config/ is already supported by some configuration loading libraries.
The most tools does not support .config/ officially (i.e. .config/ is not documented), but it is much easier to send PRs to update the documents rather than adding support for the new directory name to each tool.
Therefore, I support .config/.

**Reply by joshtriplett** [reply]
@epage Agreed. I think it makes sense to have a standard option for a visible directory, to avoid forcing files to be hidden that the project wishes to keep visible and prominent.

**Comment by Netail** [comment]
As other already pointed out, why no plural? 

Like you wouldn't use `component` as a directory name to store multiple components. Else you get this weird inconsistency between plural and singular

**Reply by joshtriplett** [reply]
@Netail Because "config" is short for "configuration", not "configurations" or "configuration files". As in, "configuration goes in here". A dozen configuration files are, collectively, the "configuration" of a project, not the "configurations" of a project.

**Reply by Netail** [reply]
Would call it `configs` honestly

**Comment by joshmeads** [comment]
I think the use of the `.config` convention comes down to a few key reasons:

### Developer Habits & Familiarity
Even though we might usually associate `~/.config` with system setups (also potentially hidden), it’s a pretty natural fit for project-level configurations too. We’ve also been using hidden files like `.namerc` for a long time without much fuss. Most modern file explorers or editors are set up to display hidden files when needed. Most devs I know keep hidden files showing at all times.

### Established Legacy and Tooling (in JS/TS)
For years, tools like [cosmiconfig](https://github.com/cosmiconfig/cosmiconfig) have embraced the .config directory approach, creating a long legacy within the JavaScript/TypeScript world. This historical support means that many projects—and by extension, developers—are already familiar with and expect that configuration can reside in this directory.

- **Cosmiconfig:**
  - [npm dependents](https://www.npmjs.com/browse/depended/cosmiconfig)
  - [GitHub dependents](https://github.com/cosmiconfig/cosmiconfig/network/dependents?dependent_type=PACKAGE)  
  - Big names like Lerna, Puppeteer, Stylelint, ShadCN, and SVGR are using this.

- **Lilconfig:**
  - [npm dependents](https://www.npmjs.com/browse/depended/lilconfig)
  - [GitHub dependents](https://github.com/antonk52/lilconfig/network/dependents?dependent_type=PACKAGE)
  - Includes cssnano, oclif, lint-staged, size-limit, playwright-test, typesync, etc

- **C12:**
  - [npm dependents](https://www.npmjs.com/browse/depended/c12)
  - [GitHub dependents](https://github.com/unjs/c12/network/dependents?dependent_type=PACKAGE)  
  - It powers many UnJS tools like Nuxt, Nitro, OpenAPI TS, and Trigger.dev.

### My personal opinion
I believe standardization is important and I do consider `.config` a defacto standard in the JS/TS community. Changing something already fairly common will lead to an even more fragmented ecosystem for years until every dependency updates.

I totally understand the perspectives being shared here—especially from folks coming from other language ecosystems where config structures might be opinionated differently. Historically the JS/TS ecosystem is by far the most config and tooling heavy—for better or worse—which is why I'm giving their existing conventions a lot of weight.

I fully agree `package.json` is not the right place. If it did have to change, I would lean more towards `_config` or `_configs` than `.configs`—simply because it's visually harder to quickly see the difference between the new plural and legacy non plural. Overall however I'd advocate for continued use of the `.config` directory.

**Reply by phanect** [reply]
Is `.config/` really a de facto standard in the JS community?
I've never heard of any tools officially supporting .config/, including those using Cosmicconfig, Lilconfig, and C12.
Even UnJS tools, which @pi0 is involved in, does not support `.config/` AFAIK.

However, if I miss many JS tools that officially support `.config/`, I support your idea.

**Reply by pi0** [reply]
Nitro, nuxt, and c12 (powered tools) have supported the `.config` dir since the beginning of this proposal.

The main initiative of using `.config/` as a dir name was consistency with XDG standards.

**Reply by phanect** [reply]
@pi0 Thanks, but I couldn't find the document page to explain that Nuxt and Nitro support `.config/`. ([Nuxt](https://nuxt.com/docs/guide/directory-structure/nuxt-config), [Nitro](https://nitro.build/guide/configuration))
Sorry for the unclearness, but *officially supported* means *documented* in this context. Without the documents, I think `.config/` is only used by few people.

If I missed the document pages about `.config/` in Nuxt and Nitro docs, that's my bad, sorry :pray:

**Reply by pi0** [reply]
PR welcome to improve docs, specially c12 🙏🏼

**Reply by phanect** [reply]
@pi0 Ah, yes, that sounds reasonable. I will send PRs to UnJS and other projects when I have time.
It might be nice to add "Documented" status to the [Support Tracker thread](https://github.com/pi0/config-dir/discussions/6).
